### ISSUE TITLE
Feature request: Go syntax deep printing

### ISSUE THREAD

**Issue by soroushjp** [davecgh/go-spew#37]
Often I am writing tests and have expected outputs from a function that are very large. I currently use a combination ginkgo/gomega `Expect` output, which is close but not Go syntax, or fmt's `%#v` printing which is Go syntax but does not dereference pointers. This would be extremely useful. If this not on the roadmap at all, I would be happy to try and implement it in go-spew.

**Comment by shurcooL** [comment]
Hey @soroushjp,

See #6. I had the _exact_ same [desire](https://github.com/shurcooL/goon#but-why) 2 years ago, and that's what my [go-goon](https://github.com/shurcooL/go-goon) fork tries to achieve.

If you do try it, let me know how it goes, and I'm interested in making it better. I use it myself very frequently.

**Comment by soroushjp** [comment]
Thanks @shurcooL, will check it out and let you know if it helps, appreciate the message!

**Comment by nathanleclaire** [comment]
:+1: on this feature request.  I'd find it immensely useful, editing the spew output by hand is not fun.

**Comment by davecgh** [comment]
I considered this during initial development, but the main issue is that there are several cases where there is no clear Go syntax to recreate a given data structure.  For example, consider the following:

```Go
type Inner struct {
	Field1 string
	Field2 **int
}
```

In this case, there is no clear Go syntax such that you can initialize an instance of Inner due to the double pointer to an int.

The same issue arises in many scenarios such as double pointers to any type, single pointers to native types, interfaces, etc.

I would certainly accept an implementation, but I suspect you'll run into some situations that are unclear how the syntax should be generated.  For example, with the provided code above, it would require the use of local variables (then naming them something reasonable is another challenge).

**Comment by nathanleclaire** [comment]
Yeah, I saw your comment in the other thread after posting.  I agree it's a difficult issue, maybe another small tool could be created for doing that in the cases where it can actually be done sanely.

**Comment by shurcooL** [comment]
FWIW, this is how go-goon currently deals with that situation:

```Go
type Inner struct {
	Field1 string
	Field2 **int
}
i := 123
ip := &i
v := Inner{
	Field1: "Text.",
	Field2: &ip,
}
goon.Dump(v)

// Output:
// (Inner)(Inner{
// 	Field1: (string)("Text."),
// 	Field2: (**int)(&&123),
// })
```

It's not valid Go code. However, IMO it's consistent and pretty readable, and only 1 line needs to be modified to make it valid Go. These situations come up less frequently, so it's still useful that the rest is dumped in Go syntax notation. It would've also been valid Go code if `Field2` value were nil.

**Comment by mitranim** [comment]
For the benefit of anyone interested, there's now a package for printing data as Go code: https://github.com/Mitranim/repr

It's made mostly for code generation, producing pretty code that you would write yourself (unlike the goon fork). 😄

There are some limitations; let me know if any critical functionality is missing.
